domain/application/state/application.go

```diff
@@ -2547,20 +2533,121 @@ ON CONFLICT(application_uuid) DO UPDATE SET
+	var (
+		ident       charmUUID
+		charmConfig charm.Config
+	)
+	if err := db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
+
+		if err := tx.Query(ctx, appStmt, appIdent).Get(&ident); errors.Is(err, sqlair.ErrNoRows) {
+			return applicationerrors.ApplicationNotFound
+		} else if err != nil {
+			return internalerrors.Capture(err)
+		}
+
+		charmConfig, err = st.getCharmConfig(ctx, tx, charmID{UUID: ident.UUID})
+		return internalerrors.Capture(err)
+	}); err != nil {
+		return "", charm.Config{}, internalerrors.Capture(err)
+	}
+
+	charmID, err := corecharm.ParseID(ident.UUID)
+	if err != nil {
+		return "", charm.Config{}, internalerrors.Errorf("parsing charm id: %w", err)
+	}
+
+	return charmID, charmConfig, nil
+}
+
+// GetApplicationIDByName returns the application ID for the named application.
+// If no application is found, an error satisfying
+// [applicationerrors.ApplicationNotFound] is returned.
+func (st *State) GetApplicationIDByName(ctx context.Context, name string) (coreapplication.ID, error) {
+	db, err := st.DB()
+	if err != nil {
+		return "", internalerrors.Capture(err)
+	}
+
+	var id coreapplication.ID
+	if err := db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
+		id, err = st.lookupApplication(ctx, tx, name)
+		return err
+	}); err != nil {
+		return "", internalerrors.Capture(err)
+	}
+	return id, nil
+}
+
+// lookupApplication looks up the application by name and returns the
+// application.ID.
+// If no application is found, an error satisfying
+// [applicationerrors.ApplicationNotFound] is returned.
+func (st *State) lookupApplication(ctx context.Context, tx *sqlair.TX, name string) (coreapplication.ID, error) {
+	app := applicationDetails{Name: name}
+	queryApplicationStmt, err := st.Prepare(`
+SELECT uuid AS &applicationDetails.uuid
+FROM application
+WHERE name = $applicationDetails.name
+`, app)
+	if err != nil {
+		return "", errors.Trace(err)
+	}
+	err = tx.Query(ctx, queryApplicationStmt, app).Get(&app)
+	if errors.Is(err, sqlair.ErrNoRows) {
+		return "", fmt.Errorf("%w: %s", applicationerrors.ApplicationNotFound, name)
+	} else if err != nil {
+		return "", errors.Annotatef(err, "looking up UUID for application %q", name)
+	}
+	return app.UUID, nil
+}
+
+func (st *State) checkApplicationCharm(ctx context.Context, tx *sqlair.TX, ident applicationID, charmID charmID) error {
+	query := `
+SELECT COUNT(*) AS &countResult.count
+FROM application
+WHERE uuid = $applicationID.uuid
+AND charm_uuid = $charmID.uuid;
+	`
+	stmt, err := st.Prepare(query, countResult{}, ident, charmID)
+	if err != nil {
+		return internalerrors.Errorf("preparing verification query: %w", err)
 	}
+
+	// Ensure that the charm is the same as the one we're trying to set.
+	var count countResult
+	if err := tx.Query(ctx, stmt, ident, charmID).Get(&count); errors.Is(err, sqlair.ErrNoRows) {
+		return applicationerrors.ApplicationHasDifferentCharm
```

Comment:
I don't think we should ever get ErrNoRows here? If the query returns no rows, `Count(*)` should return 0. So I don't think `ErrNoRows` is a handle-able error